src/scripting/lua_widget_attributes.cpp

```diff
@@ -0,0 +1,527 @@
+ * @param type: the type of the attribute, for exmaple int or std::string
+ * @param widgt_type: the type of the widget, for exmaple gui2::listbox
+ */
+
+#define WIDGET_GETTER(name, value_type, widgt_type) WIDGET_GETTER4(name, value_type, widgt_type, __LINE__)
+
+#define WIDGET_SETTER(name, value_type, widgt_type) WIDGET_SETTER4(name, value_type, widgt_type, __LINE__)
+
+
+/// CLASSIC
+
+WIDGET_GETTER("value_compat,selected_index", int, gui2::listbox)
+{
+	return w.get_selected_row() + 1;
+}
+
+/* idea
+WIDGET_GETTER("selected_widget", gui2::widget, gui2::listbox)
+{
+	if(w.get_selected_row() >= w.get_item_count()) {
+		throw std::invalid_argument("widget has no selected item");
+	}
+	return w.get_row_grid(w.get_selected_row());
+}
+*/
+
+WIDGET_SETTER("value_compat,selected_index", int, gui2::listbox)
+{
+	w.select_row(value - 1);
+}
+
+WIDGET_GETTER("value_compat,selected_index", int, gui2::multi_page)
+{
+	return w.get_selected_page() + 1;
+}
+
+WIDGET_SETTER("value_compat,selected_index", int, gui2::multi_page)
+{
+	w.select_page(value -1);
+}
+
+WIDGET_GETTER("value_compat,selected_index", int, gui2::stacked_widget)
+{
+	return w.current_layer() + 1;
+}
+
+WIDGET_SETTER("value_compat,selected_index", int, gui2::stacked_widget)
+{
+	w.select_layer(value - 1);
+}
+
+WIDGET_GETTER("selected_index", int, gui2::selectable_item)
+{
+	return w.get_value() + 1;
+}
+
+WIDGET_SETTER("selected_index", int, gui2::selectable_item)
+{
+	if(value > int(w.num_states())) {
+		throw std::invalid_argument("invalid index");
+	}
+	w.set_value(value + 1);
+}
+
+WIDGET_GETTER("value_compat,selected", bool, gui2::selectable_item)
+{
+	if(w.num_states() == 2) {
+		return w.get_value_bool();
+	}
+	throw std::invalid_argument("invalid widget");
+}
+
+WIDGET_SETTER("value_compat,selected", bool, gui2::selectable_item)
+{
+	w.set_value_bool(value);
+}
+
+WIDGET_GETTER("value_compat,text", std::string, gui2::text_box)
+{
+	return w.get_value();
+}
+
+WIDGET_SETTER("value_compat,text", std::string, gui2::text_box)
+{
+	w.set_value(value);
+}
+
+WIDGET_GETTER("value_compat,value", int, gui2::slider)
+{
+	return w.get_value();
+}
+
+WIDGET_SETTER("value_compat,value", int, gui2::slider)
+{
+	w.set_value(value);
+}
+
+WIDGET_GETTER("value_compat,percentage", int, gui2::progress_bar)
+{
+	return w.get_percentage();
+}
+
+WIDGET_SETTER("value_compat,percentage", int, gui2::progress_bar)
+{
+	w.set_percentage(value);
+}
+
+WIDGET_GETTER("value_compat,selected_item_path", std::vector<int>, gui2::tree_view)
+{
+	auto res = w.selected_item()->describe_path();
+	for(int& a : res) { ++a;}
+	return res;
+}
+
+WIDGET_GETTER("path", std::vector<int>, gui2::tree_view_node)
+{
+	auto res = w.describe_path();
+	for(int& a : res) { ++a;}
+	return res;
+}
+
+WIDGET_SETTER("value_compat,unfolded", bool, gui2::tree_view_node)
+{
+	if(value) {
+		w.unfold();
+	} else {
+		w.fold();
+	}
+}
+
+WIDGET_SETTER("value_compat,unit", lua_index_raw, gui2::unit_preview_pane)
```

Comment:
Why is there a setter for this but no getter? That seems like a pretty enormous oversight. The unit preview pane isn't all that useful if you can't even get the unit displayed!

Reply:
> Why is there a setter for this but no getter? That seems like a pretty enormous oversight. The unit preview pane isn't all that useful if you can't even get the unit displayed!

What? i think you are missing something here the unit previosu pane just shows a unit,  its a not a widget to select a unit.that is is has no user interaction, so there is a no point in such a getter.

Reply:
I would argue that there's always a point in a getter. If you have a getter, you don't need to store the unit in a separate location. If you have a list box full of unit preview panes, you don't need to keep a separate list of the units. Keeping a separate list can always risk the two representations getting out of sync if you're not careful. I'd prefer to just get the unit directly from the preview pane that was selected in the list box.

Is there a technical reason that makes it difficult to implement a getter here? If there is, then it's not the end of the world to not have one, but if there's no technical reason that makes it hard, I think it should just exist.

The same goes for other read-only widgets. For example, the text on a button should have a getter. I didn't check whether all of these _do_ have getters, maybe they do already.

Reply:
Well, then you'd need to store the 'unit' object somewhere inside the lua gui2 implementation which i still think is more work than its worth.

Reply:
Does it needed to be stored in the Lua implementation? Does the C++ widget not provide a getter already? If there's no getter in C++, then you might be right that it's more work than it's worth. I'd still prefer to have it but I can live without it.

Reply:
> Does the C++ widget not provide a getter already?

No, iirc the unit previw pane setter just sets all the child widgets to the information it extracts to the unit, meaning it does not  store the unit.

Reply:
I see. Well, I guess we can live without a getter for this attribute.